qa/suites/rbd: add test to check rbd_support module recovery

ajarr opened the pull request:
... on repeated blocklisting of its client.

There were issues with rbd_support module not being able to recover from its RADOS client being repeatedly blocklisted. This occured for example in clusters with OSDs slow to process RBD requests while the module's mirror_snapshot_scheduler was taking mirror snapshots by requesting exclusive locks on the RBD images and workloads were running on the snapshotted images via kernel clients.

Fixes: https://tracker.ceph.com/issues/62891

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

ajarr commented:
This PR depends on https://github.com/ceph/ceph/pull/53311 for the tests to run without raising POOL_APP_NOT_ENABLED error.

trociny commented on qa/workunits/rbd/rbd_support_module_recovery.sh:
```diff
@@ -0,0 +1,85 @@
+        --image ${IMAGE_PREFIX}$i 1m
+done
+
+# cleanup
```
It would be nice to wrap the cleanup thing in a `cleanup` function (which `set +e` as the first step), and set it as an exit signal handler with `trap cleanup INT TERM EXIT` (see examples in the `workunits/rbd` directory). This way if the script is interrupted for some reason it will do the cleanup anyway.

idryomov replied:
I know rbd-mirror scripts do this, but I think it's just extra complexity that often isn't implemented well.  For example, when teuthology interrupts the script due to a workunit timeout, the cluster may already be destroyed and `cleanup()` attempting to access the cluster only leads to further issues.

trociny replied:
I find it mostly useful when running such scripts locally (after adding some changes, when it is likely something will go wrong and the script will fail in the middle). Cleaning up manually after such interruption is annoying. But I will not insist. It was just a suggestion.

ajarr replied:
Thanks for the discussion. I reverted to not using `trap`

trociny commented on qa/workunits/rbd/rbd_support_module_recovery.sh:
```diff
@@ -0,0 +1,85 @@
+done
+
+# cleanup
+sudo killall fio || true
```
You might want to run `wait` after killing, to make sure we proceed only when all processes are terminated.

ajarr replied:
Sounds good. Thanks!  Added `wait` too.

ajarr commented:
Oops! Messed up by rebasing. I missed the discussion here, https://github.com/ceph/ceph/pull/53535#discussion_r1332675935 . I'll revert to not using the trap mechanism.

ajarr commented on qa/workunits/rbd/rbd_support_module_recovery.sh:
```diff
@@ -0,0 +1,85 @@
+        sleep 10
+    else
+        ceph osd blocklist add ${CLIENT_ADDR}
+	sleep 10
```
Good point! changed up the entire loop. I think this looks better.

ajarr commented on qa/workunits/rbd/rbd_support_module_recovery.sh:
```diff
@@ -0,0 +1,85 @@
+# recovers from previous blocklisting
+CURRENT_TIME=$(date +%s)
+END_TIME=$((${CURRENT_TIME}+${RUN_TIME}))
+PREV_CLIENT=0
```
OK. I missed changing the name of the variable.

ajarr commented:
I ran this through [teuthology](http://pulpito.front.sepia.ceph.com/rraja-2023-09-23_06:37:41-rbd:cli-wip-62891-distro-default-smithi/). In the 3 of 20 jobs, I could see that the [fix](https://github.com/ceph/ceph/pull/52990/commits/e66db76354093009c254aabce9e4654e27045375#diff-fa25d9e6c1abe241a10034e7c1f4821b492c28ee87e376caed0f6b4af451f494R598) for the race between librbd blocklisting and exclusive action was triggered.  The three jobs,
http://pulpito.front.sepia.ceph.com/rraja-2023-09-23_06:37:41-rbd:cli-wip-62891-distro-default-smithi/7401647/
http://pulpito.front.sepia.ceph.com/rraja-2023-09-23_06:37:41-rbd:cli-wip-62891-distro-default-smithi/7401650/
http://pulpito.front.sepia.ceph.com/rraja-2023-09-23_06:37:41-rbd:cli-wip-62891-distro-default-smithi/7401656/

In the ceph-mgr log of one of the 3 jobs, 
```
2023-09-23T07:56:12.543+0000 7f877d414700 -1 librbd::ImageWatcher: 0x558d43b6c000 error requesting lock: (108) Cannot send after transport endpoint shutdown
2023-09-23T07:56:12.543+0000 7f877d414700 20 librbd::ManagedLock: 0x558d421d4358 is_lock_owner: 0
2023-09-23T07:56:12.543+0000 7f877d414700 -1 librbd::ImageWatcher: 0x558d43b6c000 blocklisted waiting for exclusive lock
2023-09-23T07:56:12.543+0000 7f877d414700 10 librbd::ExclusiveLock: 0x558d421d4340 handle_peer_notification
2023-09-23T07:56:12.543+0000 7f877d414700 10 librbd::ManagedLock: 0x558d421d4358 send_acquire_lock:
**2023-09-23T07:56:12.543+0000 7f877d414700 -1 librbd::ManagedLock: 0x558d421d4358 send_acquire_lock: watcher not registered - client blocklisted**
2023-09-23T07:56:12.543+0000 7f877d414700 20 librbd::Operations: handle_acquire_exclusive_lock: r=-108
2023-09-23T07:56:12.543+0000 7f877d414700 15 librbd::mirror::snapshot::CreatePrimaryRequest: 0x558d43b880d0 handle_create_snapshot: r=-108
2023-09-23T07:56:12.543+0000 7f877d414700 -1 librbd::mirror::snapshot::CreatePrimaryRequest: 0x558d43b880d0 handle_create_snapshot: failed to create mirror snapshot: (108) Cannot send after transport endpoint shutdown
2023-09-23T07:56:12.543+0000 7f877d414700 15 librbd::mirror::snapshot::CreatePrimaryRequest: 0x558d43b880d0 finish: r=-108
2023-09-23T07:56:12.543+0000 7f877d414700 20 librbd: C_AioCompletion::finish: r=-108
2023-09-23T07:56:12.543+0000 7f877d414700 -1 librbd::io::AioCompletion: 0x558d437acb00 fail: (108) Cannot send after transport endpoint shutdown
2023-09-23T07:56:12.543+0000 7f877d414700 20 librbd::io::AioCompletion: 0x558d437acb00 complete_request: cb=1, pending=0
2023-09-23T07:56:12.543+0000 7f877d414700 20 librbd::io::AioCompletion: 0x558d437acb00 finalize: r=-108
2023-09-23T07:56:12.543+0000 7f877d414700  0 [rbd_support DEBUG root] CreateSnapshotRequests.handle_create_snapshot for 2//10ccc05243a3: r=-108, snap_id=None
2023-09-23T07:56:12.543+0000 7f877d414700  0 [rbd_support ERROR root] error when creating snapshot for 2//10ccc05243a3: -108
2023-09-23T07:56:12.543+0000 7f877d414700  0 [rbd_support DEBUG root] CreateSnapshotRequests.close_image 2//10ccc05243a3
```
Complete mgr log is at /a/rraja-2023-09-23_06\:37\:41-rbd\:cli-wip-62891-distro-default-smithi/7401656/remote/smithi204/log/ceph-mgr.x.log.gz in the teuthology machines.

ajarr commented on qa/workunits/rbd/rbd_support_module_recovery.sh:
```diff
@@ -0,0 +1,85 @@
+    sudo chown $(whoami) ${MNTPATHS[$i]}
+    fio --name=fiotest --filename=${MNTPATHS[$i]}/test --rw=randrw --bs=4K \
+        --direct=1 --ioengine=libaio --size=800M --iodepth=2 \
+        --runtime=${RUN_TIME} --time_based &> /dev/null &
```
Nice suggestion! I When I tried doing this, on a few occassions, I observed that sending either SIGTERM or SIGKILL to the specific fio background processes did not end up killing them for more than an hour so. Each fio background process results in a child and a grandchild process. I can acquire the list of all the descendants of each fio background process and kill them. Maybe it's okay to use `killall -e fio`  since more than one test won't run in a teuthology node?

trociny replied:
It looks like a bug in fio to me. Though knowing it does not help here.

Ideally, I would prefer if there were no assumption that the script would be running on a teuthology node only, but anyway, if `killall -e fio` works for you better then I would say just go ahead and use it. I don't think we want to complicate this.

idryomov commented on qa/workunits/rbd/rbd_support_module_recovery.sh:
```diff
@@ -0,0 +1,78 @@
+# operations on the same image.
+for ((i = 1; i <= ${NUM_IMAGES}; i++)); do
+    DEVS[$i]=$(sudo rbd device map ${POOL}/${IMAGE_PREFIX}$i)
+    sudo fio --name=fiotest --filename=${DEVS[$i]} --rw=randrw --bs=4K \
```
`sudo` shouldn't be needed -- the user on test nodes is a member of a group that allows it to access block devices.

ajarr replied:
Good to know. Thanks!

idryomov commented on qa/workunits/rbd/rbd_support_module_recovery.sh:
```diff
@@ -0,0 +1,78 @@
+for ((i = 1; i <= ${NUM_IMAGES}; i++)); do
+    DEVS[$i]=$(sudo rbd device map ${POOL}/${IMAGE_PREFIX}$i)
+    sudo fio --name=fiotest --filename=${DEVS[$i]} --rw=randrw --bs=4K \
+        --direct=1 --ioengine=libaio --size=800M --iodepth=2 \
```
Given `--runtime=43200 --time_based`, `--size` shouldn't be needed.  It only leads to questions: what is this for, why is it 800M and not 1G (the size of the image), etc.

idryomov commented on qa/workunits/rbd/rbd_support_module_recovery.sh:
```diff
@@ -0,0 +1,78 @@
+# Repeatedly blocklist rbd_support module's client ~10s after the module
+# recovers from previous blocklisting
+CURRENT_TIME=$(date +%s)
+END_TIME=$((${CURRENT_TIME}+${RUN_TIME}))
```
No need for the cluttered syntax in an arithmetic context:
```suggestion
END_TIME=$((CURRENT_TIME + RUN_TIME))
```

idryomov commented on qa/workunits/rbd/rbd_support_module_recovery.sh:
```diff
@@ -0,0 +1,78 @@
+# Run fio workloads on images via kclient
+# Test the recovery of the rbd_support module and its scheduler from their
+# librbd client being blocklisted while a exclusive lock gets passed around
+# between their librbd client and a kclient trying to perform exclusive
+# operations on the same image.
```
To be a bit more concrete:
```suggestion
# between their librbd client and a kclient trying to take mirror snapshots
# and perform I/O on the same image.
```

idryomov commented on qa/workunits/rbd/rbd_support_module_recovery.sh:
```diff
@@ -0,0 +1,78 @@
+PREV_CLIENT_ADDR=""
+CLIENT_ADDR=""
+while [[ ${CURRENT_TIME} -le ${END_TIME} ]]; do
+    if [[ ! -z "${CLIENT_ADDR}" ]] &&
```
Is this arm of the condition still needed?  Now that both variables are initialized to `""`, it seems like `"${CLIENT_ADDR}" != "${PREV_CLIENT_ADDR}"` is sufficient?

ajarr replied:
Good point!

idryomov commented on qa/workunits/rbd/rbd_support_module_recovery.sh:
```diff
@@ -0,0 +1,78 @@
+done
+
+# cleanup
+sudo killall -e fio || true
```
- once `sudo` on `fio` is dropped, can drop it here too
- given a three-character process name, does `-e` do anything?

ajarr replied:
Yes, '-e' option isn't helpful.

idryomov commented:
> At least couple of them point to issues in the recovery of rbd_support module on repeated blocklisting. e.g., http://pulpito.front.sepia.ceph.com/rraja-2023-09-23_06:37:41-rbd:cli-wip-62891-distro-default-smithi/7401648/

https://tracker.ceph.com/issues/62994

idryomov commented:
@ajarr Please file a tracker for the manager crashes in http://pulpito.front.sepia.ceph.com/rraja-2023-09-23_06:37:41-rbd:cli-wip-62891-distro-default-smithi/7401645 and http://pulpito.front.sepia.ceph.com/rraja-2023-09-23_06:37:41-rbd:cli-wip-62891-distro-default-smithi/7401646 too.

ajarr commented:
> @ajarr Please file a tracker for the manager crashes in http://pulpito.front.sepia.ceph.com/rraja-2023-09-23_06:37:41-rbd:cli-wip-62891-distro-default-smithi/7401645 and http://pulpito.front.sepia.ceph.com/rraja-2023-09-23_06:37:41-rbd:cli-wip-62891-distro-default-smithi/7401646 too.

Done.

ceph-mgr crash:  https://tracker.ceph.com/issues/63028
another hang due to exclusive lock acquisition (STATE_WAITING_FOR_LOCK) racing with blocklisting: https://tracker.ceph.com/issues/63009

ajarr commented:
> The failures I see in the [teuthology run](http://pulpito.front.sepia.ceph.com/rraja-2023-09-23_06:37:41-rbd:cli-wip-62891-distro-default-smithi/) are not related to issues in the script. 

So far filed three tracker bugs based on the failures mentioned above,
- [mgr/rbd_support: recovery from client blocklisting halts after MirrorSnapshotScheduleHandler tries to terminate its run thread](https://tracker.ceph.com/issues/62994)
- [another hang due to exclusive lock acquisition (STATE_WAITING_FOR_LOCK) racing with blocklisting](https://tracker.ceph.com/issues/63009)
- [ceph-mgr seg faults when testing for rbd_support module recovery on repeated blocklisting of its client](https://tracker.ceph.com/issues/63028)

ajarr commented:
Addressed all the comments. Ilya suggested that we wait for the bugs mentioned in https://github.com/ceph/ceph/pull/53535#issuecomment-1739671468 to be fixed before merging this test.

ajarr commented:
Observed 2 errors in this PR over multiple runs,
- See, http://qa-proxy.ceph.com/teuthology/rraja-2023-10-03_17:10:36-rbd:cli-wip-ajarr-63009-only-reacquire-kickstart-distro-default-smithi/7409690/teuthology.log
<pre>
2023-10-03T18:01:24.783 INFO:tasks.workunit.client.0.smithi031.stderr:++ ceph mgr dump
2023-10-03T18:01:24.800 INFO:tasks.workunit.client.0.smithi031.stderr:++ jq '.active_clients[]'
2023-10-03T18:01:24.800 INFO:tasks.workunit.client.0.smithi031.stderr:++ jq 'select(.name == "rbd_support")'
2023-10-03T18:01:24.801 INFO:tasks.workunit.client.0.smithi031.stderr:++ jq -r '[.addrvec[0].addr, "/", .addrvec[0].nonce|tostring] | add'
2023-10-03T18:01:25.925 INFO:tasks.workunit.client.0.smithi031.stderr:+ CLIENT_ADDR=
2023-10-03T18:01:25.925 INFO:tasks.workunit.client.0.smithi031.stderr:++ date +%s
2023-10-03T18:01:25.926 INFO:tasks.workunit.client.0.smithi031.stderr:+ CURRENT_TIME=1696356085
2023-10-03T18:01:25.926 INFO:tasks.workunit.client.0.smithi031.stderr:+ [[ 1696356085 -le 1696358112 ]]
2023-10-03T18:01:25.927 INFO:tasks.workunit.client.0.smithi031.stderr:+ [[ '' != \1\7\2\.\2\1\.\1\5\.\3\1\:\0\/\8\7\4\8\2\6\9\4\7 ]]
2023-10-03T18:01:25.927 INFO:tasks.workunit.client.0.smithi031.stderr:+ ceph osd blocklist add
2023-10-03T18:01:26.591 INFO:tasks.workunit.client.0.smithi031.stderr:Invalid command: missing required parameter addr(<EntityAddr>)
2023-10-03T18:01:26.591 INFO:tasks.workunit.client.0.smithi031.stderr:osd blocklist [<range>] <blocklistop:add|rm> <addr> [<expire:float>] :  add (optionally until <expire> seconds from now) or remove <addr> from blocklist
2023-10-03T18:01:26.591 INFO:tasks.workunit.client.0.smithi031.stderr:Error EINVAL: invalid command
</pre>
So looks like we need the check "if [[ ! -z "${CLIENT_ADDR}" ]] " mentioned in https://github.com/ceph/ceph/pull/53535#discussion_r1338392247 since sometimes `ceph mgr dump | jq .active_clients[] | jq 'select(.name == "rbd_support")' | jq -r '[.addrvec[0].addr, "/", .addrvec[0].nonce|tostring] | add' can be an empty string during repeated client blocklisting.

- See, http://qa-proxy.ceph.com/teuthology/rraja-2023-10-04_00:09:14-rbd:cli-wip-ajarr-63009-only-reacquire-kickstart-distro-default-smithi/7410093/teuthology.log,
<pre>
2023-10-04T00:30:31.931 INFO:tasks.workunit.client.0.smithi163.stderr:+ (( i <= 20 ))
2023-10-04T00:30:31.931 INFO:tasks.workunit.client.0.smithi163.stderr:++ sudo rbd device map rbd/image1
2023-10-04T00:30:32.183 INFO:tasks.workunit.client.0.smithi163.stderr:rbd: sysfs write failed
2023-10-04T00:30:32.194 INFO:tasks.workunit.client.0.smithi163.stderr:rbd: map failed: (6) No such device or address
2023-10-04T00:30:32.199 DEBUG:teuthology.orchestra.run:got remote process result: 6
2023-10-04T00:30:32.200 INFO:tasks.workunit.client.0.smithi163.stderr:+ DEVS[$i]='RBD image feature set mismatch. You can disable features unsupported by the kernel with "rbd feature disable image1 journaling".
2023-10-04T00:30:32.200 INFO:tasks.workunit.client.0.smithi163.stderr:In some cases useful info is found in syslog - try "dmesg | tail".'
2023-10-04T00:30:32.200 INFO:tasks.workunit:Stopping ['rbd/rbd_support_module_recovery.sh'] on client.0...
</pre>
Observe this when using https://github.com/ceph/ceph/blob/main/qa/suites/rbd/cli/features/journaling.yaml . Override journaling setting in rbd_support_module_recovery.sh?

idryomov commented:
> So looks like we need the check "if [[ ! -z "${CLIENT_ADDR}" ]] " mentioned in [#53535 (comment)](https://github.com/ceph/ceph/pull/53535#discussion_r1338392247) since sometimes `ceph mgr dump | jq .active_clients[] | jq 'select(.name == "rbd_support")' | jq -r '[.addrvec[0].addr, "/", .addrvec[0].nonce|tostring] | add' can be an empty string during repeated client blocklisting. I think this could be due to rbd_support module's client address getting deregistered during recovery, i.e., in MgrModule.shutdown()?

Yeah, in that case it's needed (although it can be `-n` instead of `! -z`).

> 2023-10-04T00:30:32.200 INFO:tasks.workunit.client.0.smithi163.stderr:+ DEVS[$i]='RBD image feature set mismatch. You can disable features unsupported by the kernel with "rbd feature disable image1 journaling".
> 2023-10-04T00:30:32.200 INFO:tasks.workunit.client.0.smithi163.stderr:In some cases useful info is found in syslog - try "dmesg | tail".'
> 2023-10-04T00:30:32.200 INFO:tasks.workunit:Stopping ['rbd/rbd_support_module_recovery.sh'] on client.0...
> 
> Observe this when using https://github.com/ceph/ceph/blob/main/qa/suites/rbd/cli/features/journaling.yaml . Override journaling setting in rbd_support_module_recovery.sh?

Pass `--image-feature exclusive-lock` to `rbd create`.  Alternatively, we could move this job somewhere where journaling isn't in the matrix but it's better to be explicit about needing exclusive-lock feature since this entire test revolves around it.

ajarr commented:
Thanks, Ilya! Updated the PR based on your inputs here https://github.com/ceph/ceph/pull/53535#issuecomment-1749370298

In my new test run, the journaling.yaml jobs passed, https://pulpito.ceph.com/rraja-2023-10-05_19:56:25-rbd:cli-wip-ajarr-63009-only-reacquire-kickstart-distro-default-smithi/

idryomov commented on qa/workunits/rbd/rbd_support_module_recovery.sh:
```diff
@@ -0,0 +1,77 @@
+
+# Create images and schedule their mirror snapshots
+for ((i = 1; i <= ${NUM_IMAGES}; i++)); do
+    rbd create --size 1G ${POOL}/${IMAGE_PREFIX}$i --image-feature exclusive-lock
```
Nit: `--size` has a short variant, group options before positional arguments
```suggestion
    rbd create -s 1G --image-feature exclusive-lock ${POOL}/${IMAGE_PREFIX}$i
```

idryomov commented on qa/workunits/rbd/rbd_support_module_recovery.sh:
```diff
@@ -0,0 +1,77 @@
+END_TIME=$((CURRENT_TIME + RUN_TIME))
+PREV_CLIENT_ADDR=""
+CLIENT_ADDR=""
+while [[ ${CURRENT_TIME} -le ${END_TIME} ]]; do
```
Nit: use bash arithmetic
```suggestion
while ((CURRENT_TIME <= END_TIME)); do
```

ajarr commented:
With the deadlock fix in rbd_support module (PR https://github.com/ceph/ceph/pull/54251/ ) , this PR that tests for rbd_support module recovery on repeat client blocklisting should now pass.

idryomov commented:
Yup, no issues with https://github.com/ceph/ceph/pull/54251 included (20 jobs):

https://pulpito.ceph.com/dis-2023-10-31_09:22:23-rbd-wip-dis-testing-distro-default-smithi/

idryomov commented:
Another 20-job run with https://github.com/ceph/ceph/pull/54251 included:

https://pulpito.ceph.com/dis-2023-10-31_17:19:27-rbd-wip-dis-testing-distro-default-smithi/
